### PR TITLE
feat(schematics): add custom component schematic

### LINKED ISSUE #67: Investigate component schematic applicability with lazy-loaded page structure
In Ionic 4, pages are lazy-loaded by default, but that presents some challenges when one wants to use components within pages. 

It may be best to generate an NgModule for each component so that it can be used across the app.

ref: https://blog.ionicframework.com/ionic-and-lazy-loading-pt-2/

### PR DESCRIPTION
**Pull request by mhartington**
Closes #67

### CODE REVIEW

File: schematics/component/index.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import { InsertChange } from '@schematics/angular/utility/change';
+import { buildRelativePath } from '@schematics/angular/utility/find-module';
+import { parseName } from '@schematics/angular/utility/parse-name';
+import { buildDefaultPath, getProject } from '@schematics/angular/utility/project';
+import { validateHtmlSelector, validateName } from '@schematics/angular/utility/validation';
+import * as ts from 'typescript';
+
+import { Schema as ComponentOptions } from './schema';
+
+function buildSelector(options: ComponentOptions, projectPrefix: string) {
```
Comment (dwieeb) [inline schematics/component/index.ts]: We can share `buildSelector()` in `schematics/page/index.ts`.

File: schematics/component/schema.json [hunk 2]
```diff
@@ -59,10 +59,24 @@
...
+    "createModule":  {
+      "type": "boolean",
+      "description": "Allows creating a module for the component",
+      "default": false
+    },
+    "module": {
       "type": "string",
-      "description": "Allows specification of the declaring module",
-      "alias": "m"
+      "description": "Allows adding to a modules imports or declarations"
```
Comment (dwieeb) [inline schematics/component/schema.json]: Small grammar nitpick: `modules` -> `module's`

File: schematics/component/index.ts [hunk 3]
```diff
@@ -47,8 +119,7 @@ function addImportToNgModule(options: ComponentOptions): Rule {
       }
     }
     host.commitUpdate(importRecorder);
-    return host;
```
Comment (dwieeb) [inline schematics/component/index.ts]: I presume it wasn't necessary to return host... my bad 😛

File: schematics/component/index.ts [hunk 4]
```diff
@@ -27,7 +27,79 @@ function addImportToNgModule(options: ComponentOptions): Rule {
...
+      for (const change of exportChanges) {
+        if (change instanceof InsertChange) {
+          exportRecorder.insertLeft(change.pos, change.toAdd);
+        }
+      }
+      host.commitUpdate(exportRecorder);
+    }
+
+    if (options.entryComponent) {
+      // Need to refresh the AST because we overwrote the file in the host.
```
Comment (dwieeb) [inline schematics/component/index.ts]: Can't you just commit the update later? Or does this get a bit messy?

File: schematics/component/schema.json (hunk 2 above)
Comment (mhartington) [reply schematics/component/schema.json]: 👍

File: schematics/component/index.ts (hunk 4 above)
Comment (mhartington) [reply schematics/component/index.ts]: This was copied over from the angular schematic, so Im assuming it's needed. Probably to make sure that host is properly updated and not overwriting it.

File: schematics/component/index.ts (hunk 3 above)
Comment (mhartington) [reply schematics/component/index.ts]: Host should not to return here. We return the host though further up on line 36, after the modifying the fs.

**Comment by dwieeb** [comment]
The commit message has usage examples: https://github.com/ionic-team/angular-toolkit/commit/527f54ef3659c9a752b7f82452f19bac890cf58d

**Pull request merged**